src/backend/metal/src/command.rs

```diff
@@ -2864,62 +2864,92 @@ impl com::RawCommandBuffer<Backend> for CommandBuffer {
         for (res_offset, desc_set) in pipe_layout.offsets[first_set ..].iter().zip(sets) {
             match *desc_set.borrow() {
                 native::DescriptorSet::Emulated { ref pool, ref layouts, ref sampler_range, ref texture_range, ref buffer_range } => {
-                    let data = pool.read();
                     let mut res_offset = res_offset.clone();
-                    let bind_vs = iter::once((pso::Stage::Vertex, &mut res_offset.vs, &mut self.state.resources_vs));
-                    let bind_ps = iter::once((pso::Stage::Fragment, &mut res_offset.ps, &mut self.state.resources_ps));
-
-                    for (stage, target_offset, resources) in bind_vs.into_iter().chain(bind_ps) {
-                        let mut data_offset = native::ResourceCounters {
-                            buffers: buffer_range.start as usize,
-                            textures: texture_range.start as usize,
-                            samplers: sampler_range.start as usize,
-                        };
+                    let data = pool.read();
+                    let mut data_offset = native::ResourceCounters {
+                        buffers: buffer_range.start as usize,
+                        textures: texture_range.start as usize,
+                        samplers: sampler_range.start as usize,
+                    };
 
-                        for layout in layouts.iter() {
-                            let has_stage = layout.stages.contains(stage.into());
-                            if has_stage && layout.content.contains(native::DescriptorContent::SAMPLER) {
-                                let sampler = data.samplers[data_offset.samplers];
-                                let index = target_offset.samplers;
-                                let out = &mut resources.samplers[index];
+                    for layout in layouts.iter() {
+                        //TODO: there is quite a bit code duplication happening below between vertex and fragment stages
+                        // I inlined everything as the most efficient code path I could see. Any abstraction from here
+                        // needs to ensure that the assembly code doesn't regress. Be my guess to give it a shot :)
+                        // The general idea is to only go through layouts once, and only fetch the data once.
+
+                        if layout.content.contains(native::DescriptorContent::SAMPLER) {
+                            let sampler = data.samplers[data_offset.samplers];
+                            if layout.stages.contains(pso::ShaderStageFlags::VERTEX) {
+                                let index = res_offset.vs.samplers;
```

Comment:
It would probably be easier to abstract if we could access some of these values by shader stage instead of being explicitly named `vs` etc. For example, `res_offset[stage]` and `self.state.resources[stage]`. I guess the challenge is to ensure this would be compiler-friendly enough (here and other places) so it would be unrolled to roughly the same output